### PR TITLE
Sanitise file names in multipart forms.

### LINKED ISSUE #264: Make multipart forms resilient to path traversal attacks
See https://tools.ietf.org/html/rfc7578:

> If a "filename" parameter is supplied, the requirements of
> Section 2.3 of [RFC2183] for the "receiving MUA" (i.e., the receiving
> Mail User Agent) apply to receivers of multipart/form-data as well:
> do not use the file name blindly, check and possibly change to match
> local file system conventions if applicable, and do not use directory
> path information that may be present.

### PR DESCRIPTION
**Pull request by mtrzos**
Fixes #264

### CODE REVIEW

File: safehttp/form.go [hunk 1]
```diff
@@ -212,9 +213,19 @@ func (f *MultipartForm) File(param string) []*multipart.FileHeader {
...
 		return nil
 	}
+	sanitise(fh)
 	return fh
 }
 
+// sanitise removes trailing path separators from all file names.
+// This is to ensure that uploaded files are not stored outside of the
+// designated directory.
+func sanitise(fhs []*multipart.FileHeader) {
```
Comment (kele) [inline safehttp/form.go:223]: The name is a bit too generic to be in the safehttp package.

File: safehttp/form.go [hunk 2]
```diff
@@ -212,9 +213,19 @@ func (f *MultipartForm) File(param string) []*multipart.FileHeader {
 	if !ok {
 		return nil
 	}
+	sanitise(fh)
```
Comment (kele) [inline safehttp/form.go:216]: Shouldn't this be done once when the form is parsed and returned by IncomingRequest?

**Review by kele — CHANGES_REQUESTED** [review]

File: safehttp/form.go (hunk 1 above)
Comment (mtrzos) [reply safehttp/form.go]: Channged to sanitiseFilenames

File: safehttp/form.go (hunk 2 above)
Comment (mtrzos) [reply safehttp/form.go]: moved sanitization to constructor

File: safehttp/form_test.go [hunk 3]
```diff
@@ -546,8 +544,25 @@ func TestMultipartFormValidFileAndVals(t *testing.T) {
...
+	fh := &multipart.FileHeader{Filename: "../tmp/myfile.txt"}
+	mf := &multipart.Form{
+		File: map[string][]*multipart.FileHeader{
+			"foo": {fh},
+		}}
+	f := newMulipartForm(mf)
+
+	filename := f.File("foo")[0].Filename
+	if want, got := "myfile.txt", filename; want != got {
+		t.Errorf(`f.File("foo").Filename: got %s, want %s`, got, want)
```
Comment (kele) [inline safehttp/form_test.go:557]: Use %q for strings.

File: safehttp/form_test.go [hunk 4]
```diff
@@ -546,8 +544,25 @@ func TestMultipartFormValidFileAndVals(t *testing.T) {
...
+	}
+	if err := f.Err(); err != nil {
+		t.Errorf(`f.Err(): got err %v`, err)
```
Comment (kele) [inline safehttp/form_test.go:560]: , want nil

**Review by kele — APPROVED** [review]

File: safehttp/form_test.go (hunk 3 above)
Comment (mtrzos) [reply safehttp/form_test.go]: Done

File: safehttp/form_test.go (hunk 4 above)
Comment (mtrzos) [reply safehttp/form_test.go]: Done

**Pull request merged**